File: 剑指Offer/0025.合并两个排序的链表/25.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */

function ListNode(val, next) {
  this.val = (val===undefined ? 0 : val)
  this.next = (next===undefined ? null : next)
}

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
// 方法一：用时 60ms
// 执行用时超过了 100%
// 消耗内存超过了 72%
var mergeTwoLists = function(l1, l2) {
  let result = null
  let rhead = result
  while(l1 || l2) {
    let val1 = l1 ? l1.val : 200
    let val2 = l2 ? l2.val : 200
    let curhead = val1 <= val2 ? l1 : l2
    if (!result) {
      result = curhead
      rhead = curhead
    } else {
      rhead.next = curhead
      rhead = rhead.next
    }
    val1 <= val2 ? (l1 = l1.next) : (l2 = l2.next)
  }
  return result
};

function genList(l1, l2) {
  let lnode1 = null
  let lnode2 = null
  let head = null
  for (let i = 0; i < l1.length; i++) {
    curnode = new ListNode(l1[i])
    if (!lnode1) {
      lnode1 = curnode
      head = lnode1
    } else {
      head.next = curnode
      head = head.next
    }
  }
  for (let i = 0; i < l2.length; i++) {
    curnode = new ListNode(l2[i])
    if (!lnode2) {
      lnode2 = curnode
      head = lnode2
    } else {
      head.next = curnode
      head = head.next
    }
  }
  console.log('l1: ', lnode1)
  console.log('l2: ', lnode2)
  return mergeTwoLists(lnode1, lnode2)
}

console.log(genList([1, 2, 4], [1, 3, 4]))

File: 剑指Offer/0025.合并两个排序的链表/25.go
package main

import "fmt"

// 方法一：用时 4 ms
// 执行用时超过了 96 %
// 消耗内存超过了 70 %
// 难度: 简单
func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	if l2 == nil {
		return l1
	}
	if l1 == nil {
		return l2
	}
	ret := l1
	curHead := ret
	if l2.Val < ret.Val { // 处理第一个值
		curHead = l2
		l2 = l2.Next
		curHead.Next = ret
		ret = curHead
	}
	for ; l2 != nil; {
		if curHead.Next == nil && l2.Val >= curHead.Val {
			curHead.Next = l2
			break
		}
		val1 := getVal(curHead.Next)
		val2 := getVal(l2)
		//fmt.Printf("[val1, val2] [%d, %d]\n", val1, val2)
		if val1 >= val2 {
			nextNode := curHead.Next
			curHead.Next = l2
			l2 = l2.Next
			curHead.Next.Next = nextNode
		}
		curHead = curHead.Next
	}
	return ret
}

func getVal (node *ListNode) int {
	if node != nil {
		return node.Val
	}
	return int(^uint(0) >> 1)
}

func getNextNode (node *ListNode) *ListNode {
	if node != nil {
		return node.Next
	}
	return nil
}

func main ()  {
	L1 := genList([]int{1, 1, 2})
	L2 := genList([]int{1, 2, 5, 6, 9, 9})
	ret := mergeTwoLists(L1, L2)
	fmt.Println("\n=====", ret)
	for ret != nil {
		fmt.Print(" ", ret.Val)
		ret = ret.Next
	}
	return
}

type ListNode struct {
	Val int
	Next *ListNode
}

func CreateNode(val int)  *ListNode {
	return &ListNode{val, nil}
}

func genList (trvals []int) (root *ListNode) {
	var head *ListNode
	//fmt.Println("head:", head)
	for _, val := range trvals {
		node := CreateNode(val)
		if head == nil {
			root = node
			head = root
		} else {
			head.Next = node
			head = head.Next
		}
	}
	return
}